var/spack/repos/builtin/packages/fmt/package.py

```diff
@@ -24,21 +24,54 @@ class Fmt(CMakePackage):
     version('3.0.1', sha256='4c9af0dc919a8ae7022b44e1a03c435e42d65c866f44667d8d920d342b098550')
     version('3.0.0', sha256='1b050b66fa31b74f1d75a14f15e99e728ab79572f176a53b2f8ad7c201c30ceb')
 
+    variant('cxxstd',
+            default='11',
+            values=('98', '11', '14', '17'),
+            multi=False,
+            description='Use the specified C++ standard when building')
+    variant('pic', default=True, description='Enable generation of position-independent code')
+
     depends_on('cmake@3.1.0:', type='build')
 
-    # Supported compilers are detailed here:
+    # Supported compilers/standards are detailed here:
     # http://fmtlib.net/latest/index.html#portability
     conflicts('%gcc@:4.3.999', when='@5:')
     conflicts('%llvm@:2.8.999', when='@5:')
+    # 5 and above require C++11
+    conflicts('cxxstd=98', when='@5:')
+    # 5.0.0 enables C++14 auto return types in C++11 mode
+    conflicts('cxxstd=11', when='@5.0.0')
+    # 4.1 fails with C++17 (https://github.com/fmtlib/fmt/issues/722)
+    conflicts('cxxstd=17', when='@4.1.0')
 
-    variant('pic', default=True, description='Enable generation of position-independent code')
+    # Use CMAKE_CXX_STANDARD to define C++ flag, as in later versions
+    patch('fmt-use-cmake-cxx-standard_3.0.0.patch', when='@3.0.0')
+
+    # Remove hardcoding of "-std=c++11/0x" in INTERFACE_COMPILE_OPTIONS
+    patch('fmt-no-export-cpp11flag_3.0.0.patch', when='@3.0.0:3.0.1')
+
+    # Only allow [[attributes]] on C++11 and higher
+    patch('fmt-attributes-cpp11_4.1.0.patch', when='@4.1.0')
 
     def cmake_args(self):
         spec = self.spec
         args = []
+
         if '+pic' in spec:
             args.extend([
                 '-DCMAKE_C_FLAGS={0}'.format(self.compiler.pic_flag),
                 '-DCMAKE_CXX_FLAGS={0}'.format(self.compiler.pic_flag)
             ])
+
+        args.append('-DCMAKE_CXX_STANDARD={0}'.format(
+                    spec.variants['cxxstd'].value))
```

Comment:
I've marked as a WIP because I'm now not sure whether this will handle errors correctly. `CMAKE_CXX_STANDARD` will [fallback to older standards if the requested isn't supported](https://cmake.org/cmake/help/v3.1/prop_tgt/CXX_STANDARD.html#prop_tgt:CXX_STANDARD). I could add `-DCMAKE_CXX_STANDARD_REQUIRED=ON` here to require availability of the requested standard. This would result in a failure at configure time. However, there's also the `spack.compiler.cxxXY_flag` functions which throw an exception if the compiler doesn't support the requested standard.

The returned flag wouldn't be used, but would ensure an exception from spack rather than a failure to build. What would be recommended?